### PR TITLE
bugfix: filter repeated resources and optimize memory usage.

### LINKED ISSUE #4395: `endpoints` repeated
Why `endpoints` repeated ? 
image

EnvoyGateway: v1.1.1

Here is my backend-traffic-policy:
image

### PR DESCRIPTION
**Pull request by qicz**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #4395 and optimize for #4181 

before:
image
image
image

after
image
image
image

### CODE REVIEW

**Comment by zirain** [comment]
we need some test to guratee this

**Comment by qicz** [comment]
> @qicz thanks for picking this one up, its worth investigating _why_ this happening before working around it by using a set

for my analysis, the `endpointslice` with more repeated value for the service

https://github.com/envoyproxy/gateway/blob/aab669dd99f0337afaaa7c22f7826d9ebd37cfeb/internal/provider/kubernetes/controller.go#L434-L455

logging:

image

**Comment by arkodg** [comment]
ah thanks for pointing to the controller/provider code, it looks like since the service was referenced twice, we added the corresponding endpoint slice twice. The fix in this PR is not going to work since it has visibility into a single EndpointSlice

We can either fix this in the provider or in the gateway api layer

**Comment by qicz** [comment]
> ah thanks for pointing to the controller/provider code, it looks like since the service was referenced twice, we added the corresponding endpoint slice twice. The fix in this PR is not going to work since it has visibility into a single EndpointSlice
> 
> We can either fix this in the provider or in the gateway api layer

I 'll fix this in gateway API layer

**Comment by qicz** [comment]
I tried many times but only followed resources to support duplicate resource check logic under the current implementation.

```go
var cacheKinds = []string{
	"Gateway", "ReferenceGrant", "Namespace",
	"ServiceImport", "EndpointSlice",
	"Secret", "ConfigMap", "EnvoyPatchPolicy",
	"ClientTrafficPolicy", "BackendTrafficPolicy", "SecurityPolicy",
	"BackendTLSPolicy", "EnvoyExtensionPolicy", "HTTPRouteFilter",
}
```

File: internal/gatewayapi/resource/resource_cache.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package resource
+
+// supportedCacheResourceKinds that are supported cache manager.
+// if some resources are supported remove the code slash comment.
+var supportedCacheResourceKinds = []string{
+	"EnvoyProxy",
```
Comment (shawnh2) [inline internal/gatewayapi/resource/resource_cache.go:11]: can we reuse the const from `internal/gatewayapi/resource/supported_kind.go` ?

**Review by shawnh2 — COMMENTED** [review]
thanks for the fix @qicz, now the mem usage decrease dramatically !

image

**Comment by qicz** [comment]
> thanks for the fix @qicz, now the mem usage decrease dramatically !
> 
> image

yep, we are filtering the duplicate resources that will optimize the memory for many big objects used.
cc @arkodg

**Comment by arkodg** [comment]
looks like we are already doing something similar in the provider layer using [this structure](https://github.com/envoyproxy/gateway/blob/deddeb472fa7307bf6e83f782d02ec2eba828bf1/internal/provider/kubernetes/controller.go#L160) to ensure deduplication, can we implement this there instead ?

**Comment by qicz** [comment]
> looks like we are already doing something similar in the provider layer using [this structure](https://github.com/envoyproxy/gateway/blob/deddeb472fa7307bf6e83f782d02ec2eba828bf1/internal/provider/kubernetes/controller.go#L160) to ensure deduplication, can we implement this there instead ?

They are similar,but cache used in controller and translate, IMO should in gateway layer resource pkg. resourcemap just for controller @arkodg

**Comment by arkodg** [comment]
> > looks like we are already doing something similar in the provider layer using [this structure](https://github.com/envoyproxy/gateway/blob/deddeb472fa7307bf6e83f782d02ec2eba828bf1/internal/provider/kubernetes/controller.go#L160) to ensure deduplication, can we implement this there instead ?
> 
> They are similar,but cache used in controller and translate, IMO should in gateway layer resource pkg. resourcemap just for controller @arkodg

since we already have an existing way to do this, prefer the provider approach to reduce code complexity

**Comment by qicz** [comment]
> > > looks like we are already doing something similar in the provider layer using [this structure](https://github.com/envoyproxy/gateway/blob/deddeb472fa7307bf6e83f782d02ec2eba828bf1/internal/provider/kubernetes/controller.go#L160) to ensure deduplication, can we implement this there instead ?
> > 
> > 
> > They are similar,but cache used in controller and translate, IMO should in gateway layer resource pkg. resourcemap just for controller @arkodg
> 
> since we already have an existing way to do this, prefer the provider approach to reduce code complexity

you mean reuse resourcemap struct? or reuse  resourcemap set properties? @arkodg

**Comment by qicz** [comment]
> > > > > looks like we are already doing something similar in the provider layer using [this structure](https://github.com/envoyproxy/gateway/blob/deddeb472fa7307bf6e83f782d02ec2eba828bf1/internal/provider/kubernetes/controller.go#L160) to ensure deduplication, can we implement this there instead ?
> > > > 
> > > > 
> > > > They are similar,but cache used in controller and translate, IMO should in gateway layer resource pkg. resourcemap just for controller @arkodg
> > > 
> > > 
> > > since we already have an existing way to do this, prefer the provider approach to reduce code complexity
> > 
> > 
> > you mean reuse resourcemap struct? or reuse resourcemap set properties? @arkodg
> 
> create an addition field within `resourceMappings` for service to ensure we add one unique item into the list

ok ,I'll try it but  maybe  will  occur cyle imports , translate pkg and controller cycle using.

**Comment by arkodg** [comment]
@qicz my suggestion is to solve the problem in the provider package similar to what we are doing for other resources

**Review by arkodg — APPROVED** [review]
LGTM thanks !
great to see the massive drop in mem usage

**Review by zirain — APPROVED** [review]

**Pull request merged**